osmnx/simplify.py

```diff
@@ -354,4 +371,160 @@ def clean_intersections(G, tolerance=15, dead_ends=False):
+    else:
+        # else a lot slower if scipy is not available
+        nodes_centroids = {
+            osmid : np.argmin([node['geometry'].distance(centroid) \
+                               for centroid in intersection_centroids]) \
+            for osmid, node in gdf_nodes.iterrows()
+        }
+
+    # We also build the reverse mapping: centroid -> osmids
+    centroid_elements = {i : [] for i in range(len(intersection_centroids))}
+    for i, osmid in enumerate(G.nodes(data = False)):
+        centroid = nearest_centroid[i]
+        centroid_elements[centroid].append(osmid)
+
+    # We're ready to start building the new graph
+    G__ = nx.DiGraph(name = G.graph['name'],
+                     crs = G.graph['crs'])
+
+    # Add the nodes first
+    # The centroids are given new ids = 0 .. total_centroids-1
+    for centroid, elements in centroid_elements.items():
+        x = intersection_centroids[centroid].x
+        y = intersection_centroids[centroid].y
+
+        osmid = elements[0] if len(elements) == 1 else elements
+        G__.add_node(centroid, osmid = osmid, x = x, y = y)
+
+    for centroid, elements in centroid_elements.items():
+            # get the outgoing edges for all elements in the centroid
+            out_edges = []
+            for node in elements:
+                out_edges += list(G.out_edges(node))
+
+            # select the neighbor nodes of each element in the centroid
+            out_nodes = map(lambda edge: edge[1], out_edges)
+
+            # and get the corresponding centroid of each element
+            neighbors = list(map(lambda osmid: node_centroids[osmid],
+                                 out_nodes))
+
+            # remove duplicate neighbor centroids
+            neighbors = set(neighbors)
+            # and self-loops
+            if centroid in neighbors:
+                neighbors.remove(centroid)
+
+            # helper for fixing geometries
+            centroid_point = Point(
+                intersection_centroids[centroid].x,
+                intersection_centroids[centroid].y)
+
+            # add an outgoing edge to each resulting neighbor
+            for neighbor in neighbors:
+                # retrieve the elements of this neighbor
+                neighbor_elements = centroid_elements[neighbor]
+                # select the edges (c1,c2) where c1 is an element of this centroid,
+                # and c2 is an element of this neighbor's centroid
+                out_edges_neighbor = set(filter(lambda x: x[1] in neighbor_elements, out_edges))
+
+                if len(out_edges_neighbor) > 1:
+                    # If there are multiple edges that fit this criteria, then we need
+                    # to combine their attributes, except for geometry and length
+                    # for which we need to pick a single value
+
+                    edges_attrs = [ G.edges[edge[0], edge[1], 0] for edge in out_edges_neighbor ]
+
+                    all_keys = []
```

Comment:
Can you add more comments to this code block and those that follow to document the logic?

Reply:
Addressed in [eea1bd](https://github.com/gboeing/osmnx/pull/249/commits/eea1bd636509aa4fd5fa7ec9f3dfb5c4ff12066c)